src/net/src/client_thread.cc

```diff
@@ -331,22 +331,32 @@ void ClientThread::ProcessNotifyEvents(const NetFiredEvent* pfe) {
             // connection exist
             net_multiplexer_->NetModEvent(ipport_conns_[ip_port]->fd(), 0, kReadable | kWritable);
           }
+          std::vector<std::string> msgs;
           {
             std::lock_guard l(mu_);
             auto iter = to_send_.find(ip_port);
             if (iter == to_send_.end()) {
               continue;
             }
-            // get msg from to_send_
-            std::vector<std::string>& msgs = iter->second;
-            for (auto& msg : msgs) {
-              if (ipport_conns_[ip_port]->WriteResp(msg)) {
-                to_send_[ip_port].push_back(msg);
-                NotifyWrite(ip_port);
-              }
-            }
+            msgs.swap(iter->second);
             to_send_.erase(iter);
           }
+          // get msg from to_send_
+          std::vector<std::string> send_failed_msgs;
+          for (auto& msg : msgs) {
+            if (ipport_conns_[ip_port]->WriteResp(msg)) {
+              send_failed_msgs.push_back(msg);
+              NotifyWrite(ip_port);
+            }
+          }
+          {
+            std::lock_guard l(mu_);
+            if (!send_failed_msgs.empty()) {
+              send_failed_msgs.insert(send_failed_msgs.end(), to_send_[ip_port].begin(),
+                                      to_send_[ip_port].end());
```

Comment:
```suggestion
              send_failed_msgs.insert(send_failed_msgs.end(), to_send_[ip_port].begin(), to_send_[ip_port].end());
```

And don't need check if the to_send_[ip_port] is exist? also  to_send_[ip_port]?